embrio-async/macros/src/lib.rs

```diff
@@ -110,14 +112,92 @@ pub fn async_fn(
 }
 
 fn async_fn_impl(mut item: ItemFn) -> TokenStream {
-    syn::visit_mut::visit_item_fn_mut(
-        &mut AsyncFnTransform::default(),
-        &mut item,
-    );
+    if item.asyncness.is_some() {
+        item.asyncness = None;
+        syn::visit_mut::visit_item_fn_mut(
+            &mut AsyncFnTransform::default(),
+            &mut item,
+        );
+    }
+
+    syn::visit_mut::visit_block_mut(&mut AsyncBlockTransform, &mut item.block);
+    syn::visit_mut::visit_block_mut(&mut ExpandAwait, &mut item.block);
 
     quote!(#item)
 }
 
+struct ExpandAwait;
+
+impl syn::visit_mut::VisitMut for ExpandAwait {
+    fn visit_expr_mut(&mut self, node: &mut syn::Expr) {
+        syn::visit_mut::visit_expr_mut(self, node);
+        let base = match node {
+            syn::Expr::Field(ExprField { member, base, .. }) => {
+                let member = if let Member::Named(m) = member {
+                    m
+                } else {
+                    return;
+                };
+
+                if member == "await" {
+                    &*base
+                } else {
+                    return;
+                }
+            }
+            _ => return,
+        };
+
+        *node = await_impl(base);
+    }
+}
+
+#[derive(Default)]
+struct AsyncBlockTransform;
+
+impl VisitMut for AsyncBlockTransform {
+    fn visit_expr_mut(&mut self, i: &mut Expr) {
+        syn::visit_mut::visit_expr_mut(self, i);
+        let fut = match i {
+            Expr::Async(expr_async) => {
+                if contains_yield(&expr_async.block) {
+                    async_stream_block(expr_async)
+                } else {
+                    async_block(&expr_async)
+                }
+            }
+            _ => {
+                return;
+            }
+        };
+
+        *i = fut;
+    }
+}
+
+fn contains_yield(block: &Block) -> bool {
+    struct ContainsYield(bool);
+
+    impl<'a> Visit<'a> for ContainsYield {
+        fn visit_expr_yield(&mut self, i: &'a ExprYield) {
+            syn::visit::visit_expr_yield(self, i);
```

Comment:
You don't need to continue visiting the content of the expression here. (I was trying to think of a way to make this entire visitor short-circuiting if you see a yield, but I don't think there's an easy one).